### PR TITLE
[SHELL32][SHELL32_APITEST][SDK] Implement Int64ToString

### PR DESCRIPTION
**Pull request by katahiromz**
## Purpose

Implementing missing features...
JIRA issue: N/A

## Proposed changes

- Implement `Int64ToString` and `LargeIntegerToString` functions.
- Add `Int64ToString` and `LargeIntegerToString` prototype to `<shellundoc.h>`.
- Add `Int64ToString` testcase.

## TODO

- [x] Do tests.

## Comparison

WinXP:
WinXP
Successful.

Win2k3:
Win2k3
Successful.

Win10:
Win10
Successful.

ReactOS (AFTER):
after
I found a bug of `GetNumberFormat.LeadingZero`.

### CODE REVIEW

**Comment by learn-more** [comment]
I am missing both tests and implementations for negative integers?

**Comment by katahiromz** [comment]
The negative values are supported in Vista+.

File: dll/win32/shell32/utils.cpp [hunk 1]
```diff
@@ -99,3 +99,134 @@ SHFindComputer(LPCITEMIDLIST pidlRoot, LPCITEMIDLIST pidlSavedSearch)
...
+
+    return lstrlenW(pszOut);
+}
+
+/*************************************************************************
+ *  LargeIntegerToString [SHELL32.210]
+ */
+EXTERN_C
+INT WINAPI
+LargeIntegerToString(LARGE_INTEGER *pLargeInt,
```
Comment (HBelusca) [inline dll/win32/shell32/utils.cpp:223]:
```suggestion
LargeIntegerToString(
    LARGE_INTEGER *pLargeInt,
```
Also, you can simplify this with `PLARGE_INTEGER`.

In addition you can restyle similarly the parameters list of the other new functions you've added in this file. What about also adding SAL annotations?

File: modules/rostests/apitests/shell32/Int64ToString.cpp [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+    format.lpThousandSep = szThousandSep;
+    format.NegativeOrder = 0;
+
+    ret = Int64ToString(10000000000, szBuff, _countof(szBuff), FALSE, &format, -1);
+    ok_int(ret, 11);
+    ok_wstr(szBuff, L"10000000000");
+
+    ret = Int64ToString(10000000000, szBuff, _countof(szBuff), TRUE, &format, -1);
+    ok_int(ret, 17);
+    ok_wstr(szBuff, L"10,000,000,000.00");
```
Comment (HBelusca) [inline modules/rostests/apitests/shell32/Int64ToString.cpp:61]: Isn't this depending on the ambient per-user regional number formatting?

File: dll/win32/shell32/utils.cpp [hunk 3]
```diff
@@ -99,3 +99,134 @@ SHFindComputer(LPCITEMIDLIST pidlRoot, LPCITEMIDLIST pidlSavedSearch)
 
     return SUCCEEDED(hr);
 }
+
+static VOID Int64ToStr(LONGLONG llValue, LPWSTR pszValue)
+{
+    WCHAR szBuff[40];
+    INT ich = 0;
+#if (WINVER >= 0x0600)
```
Comment (HBelusca) [inline dll/win32/shell32/utils.cpp:107]:
```suggestion
#if (_WIN32_WINNT >= _WIN32_WINNT_VISTA)
```

File: dll/win32/shell32/utils.cpp [hunk 4]
```diff
@@ -99,3 +99,134 @@ SHFindComputer(LPCITEMIDLIST pidlRoot, LPCITEMIDLIST pidlSavedSearch)
...
+        llValue = -llValue;
+#endif
+
+    do
+    {
+        szBuff[ich++] = (WCHAR)(L'0' + (llValue % 10));
+        llValue /= 10;
+    } while (llValue != 0);
+
+#if (WINVER >= 0x0600)
```
Comment (HBelusca) [inline dll/win32/shell32/utils.cpp:120]:
```suggestion
#if (_WIN32_WINNT >= _WIN32_WINNT_VISTA)
```

File: dll/win32/shell32/utils.cpp (hunk 3 above)
Comment (JoachimHenze) [reply dll/win32/shell32/utils.cpp:107]: Can we unhack it in the CMakeList.txt? Like we/you recently did in some other module?

File: modules/rostests/apitests/shell32/Int64ToString.cpp (hunk 2 above)
Comment (katahiromz) [reply modules/rostests/apitests/shell32/Int64ToString.cpp:61]: `&format` is specified, and `bUseFormat` is `TRUE`, and the flags is `-1`. So the user preference doesn't affect this behavior.

File: dll/win32/shell32/utils.cpp (hunk 1 above)
Comment (katahiromz) [reply dll/win32/shell32/utils.cpp]: Added SAL.
Fixed prototypes (I added some `const`).

File: dll/win32/shell32/utils.cpp (hunk 4 above)
Comment (katahiromz) [reply dll/win32/shell32/utils.cpp:120]: Non, non. `_WIN32_WINNT` is hacked at here. See `CMakeLists.txt`.

File: dll/win32/shell32/utils.cpp [hunk 5]
```diff
@@ -99,3 +99,135 @@ SHFindComputer(LPCITEMIDLIST pidlRoot, LPCITEMIDLIST pidlSavedSearch)
...
+    _In_ UINT cchOut,
+    _In_ BOOL bUseFormat,
+    _In_ const NUMBERFMTW *pNumberFormat,
+    _In_ DWORD dwNumberFlags)
+{
+    INT ret;
+    NUMBERFMTW NumFormat;
+    WCHAR szValue[80], szDecimalSep[MAX_NUMSEP], szThousandSep[MAX_NUMSEP];
+
+    Int64ToStr(llValue, szValue);
```
Comment (HBelusca) [inline dll/win32/shell32/utils.cpp:201]: It would be better if the internal Int64ToStr function could take the size of the szValue buffer, in order to make it clear it can handle a potential ill-sized buffer.

File: dll/win32/shell32/utils.cpp [hunk 6]
```diff
@@ -99,3 +99,135 @@ SHFindComputer(LPCITEMIDLIST pidlRoot, LPCITEMIDLIST pidlSavedSearch)
...
+    } while (ich > 0);
+
+    *pszValue = UNICODE_NULL;
+}
+
+#define MAX_NUMSEP 6
+
+static VOID
+Int64GetNumFormat(NUMBERFMTW *pDest, const NUMBERFMTW *pSrc, DWORD dwNumberFlags,
+                  LPWSTR pszDecimal, LPWSTR pszThousand)
```
Comment (HBelusca) [inline dll/win32/shell32/utils.cpp:138]: Please add SAL annotations, and provide explicit parameters for the sizes of these pszDecimal and pszThousand buffers, instead of hardcoding them below in that function.

File: dll/win32/shell32/utils.cpp [hunk 7]
```diff
@@ -99,3 +99,135 @@ SHFindComputer(LPCITEMIDLIST pidlRoot, LPCITEMIDLIST pidlSavedSearch)
...
+{
+    INT ret;
+    NUMBERFMTW NumFormat;
+    WCHAR szValue[80], szDecimalSep[MAX_NUMSEP], szThousandSep[MAX_NUMSEP];
+
+    Int64ToStr(llValue, szValue);
+
+    if (bUseFormat)
+    {
+        Int64GetNumFormat(&NumFormat, pNumberFormat, dwNumberFlags, szDecimalSep, szThousandSep);
```
Comment (HBelusca) [inline dll/win32/shell32/utils.cpp:205]: The function need to accept the actual sizes of the szDecimal and ThousandSep buffers.

**Comment by HBelusca** [comment]
I'm not sure the idea of "unhacking" the WIN32_WINNT thing was a good idea, actually.

File: dll/win32/shell32/utils.cpp (hunk 5 above)
Comment (katahiromz) [reply dll/win32/shell32/utils.cpp]: Fixed.

File: dll/win32/shell32/utils.cpp (hunk 6 above)
Comment (katahiromz) [reply dll/win32/shell32/utils.cpp]: SAL annotations added.

File: dll/win32/shell32/utils.cpp (hunk 7 above)
Comment (katahiromz) [reply dll/win32/shell32/utils.cpp]: Fixed.

**Comment by katahiromz** [comment]
VirtualBox_ReactOS_20_09_2023_21_01_02
This PR is still valid. We have to decide whether to unhack or not.

File: dll/win32/shell32/wine/shpolicy.c [hunk 8]
```diff
@@ -70,7 +70,9 @@ static const char strNoChangeStartMenu[] = {"NoChangeStartMenu"};
 static const char strNoWindowsUpdate[] = {"NoWindowsUpdate"};
 static const char strNoSetActiveDesktop[] = {"NoSetActiveDesktop"};
 static const char strNoForgetSoftwareUpdate[] = {"NoForgetSoftwareUpdate"};
-//static const char strNoMSAppLogo[] = {"NoMSAppLogo5ChannelNotify"};
+#if (NTDDI_VERSION < NTDDI_LONGHORN)
+static const char strNoMSAppLogo[] = {"NoMSAppLogo5ChannelNotify"};
```
Comment (katahiromz) [inline dll/win32/shell32/wine/shpolicy.c:74]: What do you think about this change?

**Comment by katahiromz** [comment]
We have to normalize the version control. I am going to unhack the `_WIN32_WINNT` hack.

**Comment by HBelusca** [comment]
> We have to normalize the version control. I am going to unhack the `_WIN32_WINNT` hack.

But then please do it in a separate PR. And this so-called "hack" way of doing things is used in many other places than the shell (for good reasons sometimes).

**Comment by katahiromz** [comment]
Unhacking goes #5708.

**Comment by HBelusca** [comment]
> Unhacking goes #5708.

The present Int64ToString PR may be reset to open state since using WINVER may be OK for the time being.

File: dll/win32/shell32/wine/shpolicy.c (hunk 8 above)
Comment (HBelusca) [reply dll/win32/shell32/wine/shpolicy.c:74]: Unrelated to the Int64ToString PR.

File: dll/win32/shell32/utils.cpp [hunk 9]
```diff
@@ -99,3 +99,151 @@ SHFindComputer(LPCITEMIDLIST pidlRoot, LPCITEMIDLIST pidlSavedSearch)
 
     return SUCCEEDED(hr);
 }
+
+static HRESULT Int64ToStr(LONGLONG llValue, LPWSTR pszValue, UINT cchValue)
+{
+    WCHAR szBuff[40];
+    UINT ich = 0, ichValue;
+#if (WINVER >= _WIN32_WINNT_VISTA)
```
Comment (HBelusca) [inline dll/win32/shell32/utils.cpp:107]: ok with WINVER, prefer using e.g. WINVER_VISTA (OK I know the values usually match with the WIN32_WINNT_xxx defines...)

File: dll/win32/shell32/utils.cpp [hunk 10]
```diff
@@ -99,3 +99,151 @@ SHFindComputer(LPCITEMIDLIST pidlRoot, LPCITEMIDLIST pidlSavedSearch)
 
     return SUCCEEDED(hr);
 }
+
+static HRESULT Int64ToStr(LONGLONG llValue, LPWSTR pszValue, UINT cchValue)
```
Comment (HBelusca) [inline dll/win32/shell32/utils.cpp:103]:
```suggestion
static HRESULT
Int64ToStr(
    _In_ LONGLONG llValue,
    _Out_writes_z_(cchValue) LPWSTR pszValue,
    _In_ UINT cchValue)
```

File: dll/win32/shell32/utils.cpp [hunk 11]
```diff
@@ -99,3 +99,151 @@ SHFindComputer(LPCITEMIDLIST pidlRoot, LPCITEMIDLIST pidlSavedSearch)
...
+    pszValue[ichValue] = UNICODE_NULL;
+    return S_OK;
+}
+
+static VOID
+Int64GetNumFormat(
+    _Out_ NUMBERFMTW *pDest,
+    _In_opt_ const NUMBERFMTW *pSrc,
+    _In_ DWORD dwNumberFlags,
+    _Out_writes_(cchDecimal) LPWSTR pszDecimal,
```
Comment (HBelusca) [inline dll/win32/shell32/utils.cpp:149]:
```suggestion
    _Out_writes_z_(cchDecimal) LPWSTR pszDecimal,
```

File: dll/win32/shell32/utils.cpp [hunk 12]
```diff
@@ -99,3 +99,151 @@ SHFindComputer(LPCITEMIDLIST pidlRoot, LPCITEMIDLIST pidlSavedSearch)
...
+    _In_ INT cchDecimal,
+    _Out_writes_(cchThousand) LPWSTR pszThousand,
```
Comment (HBelusca) [inline dll/win32/shell32/utils.cpp:151]:
```suggestion
    _Out_writes_z_(cchThousand) LPWSTR pszThousand,
```

File: dll/win32/shell32/utils.cpp [hunk 13]
```diff
@@ -99,3 +99,151 @@ SHFindComputer(LPCITEMIDLIST pidlRoot, LPCITEMIDLIST pidlSavedSearch)
...
+    return S_OK;
+}
+
+static VOID
+Int64GetNumFormat(
+    _Out_ NUMBERFMTW *pDest,
+    _In_opt_ const NUMBERFMTW *pSrc,
+    _In_ DWORD dwNumberFlags,
+    _Out_writes_(cchDecimal) LPWSTR pszDecimal,
+    _In_ INT cchDecimal,
```
Comment (HBelusca) [inline dll/win32/shell32/utils.cpp:150]: Technically should be unsigned int (UINT), but yes, I know that GetLocaleInfoW takes the buffer size in (signed) INT, for whatever reason...

File: dll/win32/shell32/utils.cpp [hunk 14]
```diff
@@ -99,3 +99,151 @@ SHFindComputer(LPCITEMIDLIST pidlRoot, LPCITEMIDLIST pidlSavedSearch)
...
+}
+
+/*************************************************************************
+ *  LargeIntegerToString [SHELL32.210]
+ */
+EXTERN_C
+INT WINAPI
+LargeIntegerToString(
+    _In_ const LARGE_INTEGER *pLargeInt,
+    _Out_writes_(cchOut) LPWSTR pszOut,
```
Comment (HBelusca) [inline dll/win32/shell32/utils.cpp:241]:
```suggestion
    _Out_writes_z_(cchOut) LPWSTR pszOut,
```

File: sdk/include/reactos/undocshell.h [hunk 15]
```diff
@@ -658,6 +658,32 @@ BOOL WINAPI GUIDFromStringW(
...
+    _Out_writes_(cchOut) LPWSTR pszOut,
+    _In_ UINT cchOut,
+    _In_ BOOL bUseFormat,
+    _In_opt_ const NUMBERFMTW *pNumberFormat,
+    _In_ DWORD dwNumberFlags);
+
+INT WINAPI
+LargeIntegerToString(
+    _In_ const LARGE_INTEGER *pLargeInt,
+    _Out_writes_(cchOut) LPWSTR pszOut,
```
Comment (HBelusca) [inline sdk/include/reactos/undocshell.h:681]:
```suggestion
    _Out_writes_z_(cchOut) LPWSTR pszOut,
```

File: sdk/include/reactos/undocshell.h [hunk 16]
```diff
@@ -658,6 +658,32 @@ BOOL WINAPI GUIDFromStringW(
...
+#define NUMFLAG_LEADINGZERO    0x02
+#define NUMFLAG_GROUPING       0x04
+#define NUMFLAG_DECIMALSEP     0x08
+#define NUMFLAG_THOUSANDSEP    0x10
+#define NUMFLAG_NEGATIVEORDER  0x20
+
+INT WINAPI
+Int64ToString(
+    _In_ LONGLONG llValue,
+    _Out_writes_(cchOut) LPWSTR pszOut,
```
Comment (HBelusca) [inline sdk/include/reactos/undocshell.h:672]:
```suggestion
    _Out_writes_z_(cchOut) LPWSTR pszOut,
```

File: dll/win32/shell32/utils.cpp [hunk 17]
```diff
@@ -99,3 +99,151 @@ SHFindComputer(LPCITEMIDLIST pidlRoot, LPCITEMIDLIST pidlSavedSearch)
...
+}
+
+/*************************************************************************
+ *  Int64ToString [SHELL32.209]
+ */
+EXTERN_C
+INT WINAPI
+Int64ToString(
+    _In_ LONGLONG llValue,
+    _Out_writes_(cchOut) LPWSTR pszOut,
```
Comment (HBelusca) [inline dll/win32/shell32/utils.cpp:205]:
```suggestion
    _Out_writes_z_(cchOut) LPWSTR pszOut,
```

**Review by HBelusca — COMMENTED** [review]
Minor stuff but otherwise LGTM.
Maybe you could mention those two articles either in the commit message or as part of function "documentation" -- in those little comments that precede the functions:
http://undoc.airesoft.co.uk/shell32.dll/Int64ToString.php
http://undoc.airesoft.co.uk/shell32.dll/LargeIntegerToString.php

**Review by HBelusca — APPROVED** [review]

**Comment by wjk** [comment]
Unless I am missing something, `NUMFLAG_LEADINGZERO` is not handled, nor is there a test for its use.

**Comment by katahiromz** [comment]
`LeadingZero` is now covered.
